src/Cake.Common/Build/MyGet/MyGetProvider.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using Cake.Core;
+
+namespace Cake.Common.Build.MyGet
+{
+    /// <summary>
+    /// Responsible for communicating with MyGet.
+    /// </summary>
+    public sealed class MyGetProvider : IMyGetProvider
+    {
+        private const string MessagePrefix = "##myget[";
+        private const string MessagePostfix = "]";
+        private static readonly Dictionary<string, string> _sanitizationTokens;
+        private readonly ICakeEnvironment _environment;
+
+        static MyGetProvider()
+        {
+            _sanitizationTokens = new Dictionary<string, string>
+            {
+                { "|", "||" },
+                { "\'", "|\'" },
+                { "\n", "|n" },
+                { "\r", "|r" },
+                { "[", "|['" },
+                { "]", "|]" }
+            };
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MyGetProvider"/> class.
+        /// </summary>
+        /// <param name="environment">The cake environment.</param>
+        public MyGetProvider(ICakeEnvironment environment)
+        {
+            if (environment == null)
+            {
+                throw new ArgumentNullException("environment");
+            }
+
+            _environment = environment;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the current build is running on MyGet.
+        /// </summary>
+        /// <value>
+        /// <c>true</c> if the current build is running on MyGet; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsRunningOnMyGet
+        {
+            get { return !string.IsNullOrWhiteSpace(_environment.GetEnvironmentVariable("BuildRunner")); }
+        }
+
+        /// <summary>
+        /// Report a build problem to MyGet.
+        /// </summary>
+        /// <param name="description">Description of build problem.</param>
+        public void BuildProblem(string description)
+        {
+            WriteServiceMessage("buildProblem", "description", description);
+        }
+
+        /// <summary>
+        /// Allows setting an environment variable that can be used by a future process.
+        /// </summary>
+        /// <param name="name">Name of the parameter to set.</param>
+        /// <param name="value">Value to assign to the parameter.</param>
+        public void SetParameter(string name, string value)
+        {
+            WriteServiceMessage("setParameter", new Dictionary<string, string>
+            {
+                { "name", name },
+                { "value", value }
+            });
+        }
+
+        /// <summary>
+        /// Write a status message to the MyGet build log.
+        /// </summary>
+        /// <param name="message">Message contents.</param>
+        /// <param name="status">Build status.</param>
```

Comment:
Maybe you can introduce an enum for this instead of magic strings.

```csharp
public enum MyGetStatus
{
   Failure,
   Error,
   Warning,
   Normal
}
```

Or map the `Verbose` enum to the MyGet status codes.